source/extensions/filters/network/redis_proxy/router_impl.cc

```diff
@@ -0,0 +1,80 @@
+#include "extensions/filters/network/redis_proxy/router_impl.h"
+
+#include <algorithm>
+
+#include "extensions/filters/network/redis_proxy/conn_pool_impl.h"
+
+namespace Envoy {
+namespace Extensions {
+namespace NetworkFilters {
+namespace RedisProxy {
+
+InstanceFactoryImpl::InstanceFactoryImpl(Upstream::ClusterManager& cm,
+                                         ThreadLocal::SlotAllocator& tls)
+    : cm_(cm), tls_(tls), cf_(ConnPool::ClientFactoryImpl::instance_) {}
+
+ConnPool::InstancePtr InstanceFactoryImpl::create(
+    const envoy::config::filter::network::redis_proxy::v2::RedisProxy::ForwardAction& action) {
+  return std::make_unique<ConnPool::InstanceImpl>(action.cluster(), cm_, cf_, tls_,
+                                                  action.settings());
+}
+
+RouterImpl::RouterImpl(
+    const envoy::config::filter::network::redis_proxy::v2::RedisProxy::RouteConfig& route_config,
+    InstanceFactory& factory)
+    : factory_(factory), prefix_routes_(createPrefixRoutes(route_config)),
+      catch_all_(route_config.has_catch_all() ? factory.create(route_config.catch_all())
+                                              : nullptr) {}
+
+RouterImpl::PrefixRouteSet RouterImpl::createPrefixRoutes(
+    const envoy::config::filter::network::redis_proxy::v2::RedisProxy::RouteConfig& route_config) {
+
+  RouterImpl::PrefixRouteSet routes;
+  for (auto const route : route_config.routes()) {
+    auto conn_pool = factory_.create(route.route());
+    auto match = route.match();
+
+    routes.emplace(std::move(conn_pool), match.prefix(), match.case_insensitive());
+  }
+  return routes;
+}
+
+ConnPool::PoolRequest* RouterImpl::makeRequest(const std::string& key, const RespValue& request,
+                                               ConnPool::PoolCallbacks& callbacks) {
+
+  for (auto const& route : prefix_routes_) {
```

Comment:
Can you add some comments here about the match semantics? AFAICT this is an iterative search over an ordered set, with longest match wins? Meaning, per our other discussion this is not going to give O(1) performance, it will still be an O(N) match, right? Can we document that fact in the public docs also?